File: web/bone_colors.js
/**
 * Bone color configuration
 * Each bone can have its color customized here
 * Format: "bone_name" or ["joint1", "joint2"] -> hex color code
 * 
 * OpenPose BODY_18 limbSeq (1-indexed in reference, converted to 0-indexed joint names):
 * [[2,3], [2,6], [3,4], [4,5], [6,7], [7,8], [2,9], [9,10], [10,11], [2,12], [12,13], [13,14], [2,1], [1,15], [15,17], [1,16], [16,18], [3,17], [6,18]]
 * Maps to:
 * 0: neck->r_shoulder, 1: neck->l_shoulder, 2: r_shoulder->r_elbow, 3: r_elbow->r_wrist, 
 * 4: l_shoulder->l_elbow, 5: l_elbow->l_wrist, 6: neck->r_hip, 7: r_hip->r_knee, 8: r_knee->r_ankle,
 * 9: neck->l_hip, 10: l_hip->l_knee, 11: l_knee->l_ankle, 12: neck->nose, 13: nose->r_eye,
 * 14: r_eye->r_ear, 15: nose->l_eye, 16: l_eye->l_ear, 17: r_shoulder->r_ear, 18: l_shoulder->l_ear
 */

// OpenPose standard color palette (18 colors from reference implementation)
const OPENPOSE_COLORS = [
    "#ff0000",  // 0: Red
    "#ff5500",  // 1: Orange
    "#ffaa00",  // 2: Dark orange
    "#ffff00",  // 3: Yellow
    "#aaff00",  // 4: Yellow-green
    "#55ff00",  // 5: Light green
    "#00ff00",  // 6: Green
    "#00ff55",  // 7: Green-cyan
    "#00ffaa",  // 8: Cyan-green
    "#00ffff",  // 9: Cyan
    "#00aaff",  // 10: Cyan-blue
    "#0055ff",  // 11: Light blue
    "#0000ff",  // 12: Blue
    "#5500ff",  // 13: Purple-blue
    "#aa00ff",  // 14: Purple
    "#ff00ff",  // 15: Magenta
    "#ff00aa",  // 16: Pink
    "#ff0055",  // 17: Hot pink
];

export const BONE_COLORS = {
    // Upper body
    "nose-neck": OPENPOSE_COLORS[12],             // 0: Blue
    "neck-r_shoulder": OPENPOSE_COLORS[1],        // 1: Orange - RIGHT is orange/warm
    "r_shoulder-r_elbow": OPENPOSE_COLORS[2],     // 2: Dark orange - RIGHT
    "r_elbow-r_wrist": OPENPOSE_COLORS[3],        // 3: Yellow - RIGHT
    "neck-l_shoulder": OPENPOSE_COLORS[5],        // 4: Light green - LEFT is cool
    "l_shoulder-l_elbow": OPENPOSE_COLORS[6],     // 5: Green - LEFT
    "l_elbow-l_wrist": OPENPOSE_COLORS[7],        // 6: Green-cyan - LEFT
    "neck-r_hip": OPENPOSE_COLORS[8],             // 7: Cyan-green - RIGHT hip
    "neck-l_hip": OPENPOSE_COLORS[11],            // 8: Light blue - LEFT hip

    // Right leg (GREEN)
    "r_hip-r_knee": OPENPOSE_COLORS[5],           // 9: Light green - RIGHT leg
    "r_knee-r_ankle": OPENPOSE_COLORS[6],         // 10: Green - RIGHT leg

    // Left leg (CYAN-BLUE)
    "l_hip-l_knee": OPENPOSE_COLORS[9],           // 11: Cyan - LEFT leg
    "l_knee-l_ankle": OPENPOSE_COLORS[10],        // 12: Cyan-blue - LEFT leg

    // Face
    "nose-r_eye": OPENPOSE_COLORS[14],            // 13: Purple
    "r_eye-r_ear": OPENPOSE_COLORS[16],           // 14: Pink
    "nose-l_eye": OPENPOSE_COLORS[14],            // 15: Purple
    "l_eye-l_ear": OPENPOSE_COLORS[16],           // 16: Pink
};

// Fallback palette - MUST match Python FALLBACK_PALETTE exactly
export const FALLBACK_PALETTE = [
    OPENPOSE_COLORS[12],  // 0: nose->neck (Blue)
    OPENPOSE_COLORS[1],   // 1: neck->r_shoulder (Orange) - RIGHT is orange/warm
    OPENPOSE_COLORS[2],   // 2: r_shoulder->r_elbow (Dark orange) - RIGHT
    OPENPOSE_COLORS[3],   // 3: r_elbow->r_wrist (Yellow) - RIGHT
    OPENPOSE_COLORS[5],   // 4: neck->l_shoulder (Light green) - LEFT is cool
    OPENPOSE_COLORS[6],   // 5: l_shoulder->l_elbow (Green) - LEFT
    OPENPOSE_COLORS[7],   // 6: l_elbow->l_wrist (Green-cyan) - LEFT
    OPENPOSE_COLORS[8],   // 7: neck->r_hip (Cyan-green) - RIGHT hip
    OPENPOSE_COLORS[11],  // 8: neck->l_hip (Light blue) - LEFT hip
    OPENPOSE_COLORS[5],   // 9: r_hip->r_knee (Light green) - RIGHT leg
    OPENPOSE_COLORS[6],   // 10: r_knee->r_ankle (Green) - RIGHT leg
    OPENPOSE_COLORS[9],   // 11: l_hip->l_knee (Cyan) - LEFT leg
    OPENPOSE_COLORS[10],  // 12: l_knee->l_ankle (Cyan-blue) - LEFT leg
    OPENPOSE_COLORS[14],  // 13: nose->r_eye (Purple)
    OPENPOSE_COLORS[16],  // 14: r_eye->r_ear (Pink)
    OPENPOSE_COLORS[14],  // 15: nose->l_eye (Purple)
    OPENPOSE_COLORS[16],  // 16: l_eye->l_ear (Pink)
];

/**
 * Get color for a specific bone connection
 * @param {string} joint1 - First joint name
 * @param {string} joint2 - Second joint name
 * @param {number} boneIndex - Fallback bone index if direct lookup fails
 * @returns {string} Hex color code
 */
export function getBoneColor(joint1, joint2, boneIndex) {
    // Primary: use boneIndex for consistent palette-based colors
    if (boneIndex >= 0 && boneIndex < FALLBACK_PALETTE.length) {
        return FALLBACK_PALETTE[boneIndex];
    }
    
    // Secondary: try direct lookup
    const key1 = `${joint1}-${joint2}`;
    if (BONE_COLORS[key1]) {
        return BONE_COLORS[key1];
    }
    
    // Tertiary: try reverse lookup
    const key2 = `${joint2}-${joint1}`;
    if (BONE_COLORS[key2]) {
        return BONE_COLORS[key2];
    }
    
    return "#ffffff";
}

/**
 * Convert hex color to RGB
 * @param {string} hex - Hex color code (e.g., "#ff0000")
 * @returns {object} Object with r, g, b properties (0-255)
 */
export function hexToRgb(hex) {
    const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
    return result ? {
        r: parseInt(result[1], 16),
        g: parseInt(result[2], 16),
        b: parseInt(result[3], 16)
    } : { r: 255, g: 255, b: 255 };
}

/**
 * Convert hex color to RGBA string for canvas
 * @param {string} hex - Hex color code
 * @param {number} alpha - Alpha value (0-1)
 * @returns {string} RGBA string
 */
export function hexToRgba(hex, alpha = 1) {
    const rgb = hexToRgb(hex);
    return `rgba(${rgb.r}, ${rgb.g}, ${rgb.b}, ${alpha})`;
}

/**
 * Convert hex color to BGR tuple for OpenCV
 * @param {string} hex - Hex color code
 * @returns {array} [B, G, R] array for OpenCV
 */
export function hexToBgr(hex) {
    const rgb = hexToRgb(hex);
    return [rgb.b, rgb.g, rgb.r];
}

File: pose_utils/bone_colors.py
"""Bone color configuration for Python side"""

# Color palette shared between web widget and Python renderer.
# Ordering matches `BONE_CONNECTIONS` defined in skeleton_512x1536.py and
# the JavaScript editor so preview and export use identical colors.

# OpenPose standard color palette (18 colors from reference implementation)
# RGB format for Python/OpenCV (note: OpenCV uses BGR, so colors are in RGB here and converted when drawing)
OPENPOSE_COLORS = [
    (255, 0, 0),      # 0: Red
    (255, 85, 0),     # 1: Orange
    (255, 170, 0),    # 2: Dark orange
    (255, 255, 0),    # 3: Yellow
    (170, 255, 0),    # 4: Yellow-green
    (85, 255, 0),     # 5: Light green
    (0, 255, 0),      # 6: Green
    (0, 255, 85),     # 7: Green-cyan
    (0, 255, 170),    # 8: Cyan-green
    (0, 255, 255),    # 9: Cyan
    (0, 170, 255),    # 10: Cyan-blue
    (0, 85, 255),     # 11: Light blue
    (0, 0, 255),      # 12: Blue
    (85, 0, 255),     # 13: Purple-blue
    (170, 0, 255),    # 14: Purple
    (255, 0, 255),    # 15: Magenta
    (255, 0, 170),    # 16: Pink
    (255, 0, 85),     # 17: Hot pink
]

# FALLBACK_PALETTE is used when direct name lookup fails.
# It maps bone index (position in BONE_CONNECTIONS) to color index in OPENPOSE_COLORS.
# MUST match JavaScript FALLBACK_PALETTE exactly.
FALLBACK_PALETTE = [
    OPENPOSE_COLORS[12],  # 0: nose->neck (Blue)
    OPENPOSE_COLORS[5],   # 1: neck->r_shoulder (Light green) - SWAPPED
    OPENPOSE_COLORS[6],   # 2: r_shoulder->r_elbow (Green) - SWAPPED
    OPENPOSE_COLORS[7],   # 3: r_elbow->r_wrist (Green-cyan) - SWAPPED
    OPENPOSE_COLORS[1],   # 4: neck->l_shoulder (Orange) - SWAPPED
    OPENPOSE_COLORS[2],   # 5: l_shoulder->l_elbow (Dark orange) - SWAPPED
    OPENPOSE_COLORS[3],   # 6: l_elbow->l_wrist (Yellow) - SWAPPED
    OPENPOSE_COLORS[11],  # 7: neck->r_hip (Light blue) - SWAPPED
    OPENPOSE_COLORS[8],   # 8: neck->l_hip (Cyan-green) - SWAPPED
    OPENPOSE_COLORS[12],  # 9: r_hip->r_knee (Blue) - SWAPPED
    OPENPOSE_COLORS[13],  # 10: r_knee->r_ankle (Purple-blue) - SWAPPED
    OPENPOSE_COLORS[9],   # 11: l_hip->l_knee (Cyan) - SWAPPED
    OPENPOSE_COLORS[10],  # 12: l_knee->l_ankle (Cyan-blue) - SWAPPED
    OPENPOSE_COLORS[14],  # 13: nose->r_eye (Purple)
    OPENPOSE_COLORS[16],  # 14: r_eye->r_ear (Pink)
    OPENPOSE_COLORS[14],  # 15: nose->l_eye (Purple)
    OPENPOSE_COLORS[16],  # 16: l_eye->l_ear (Pink)
]

# Color palette for bones (matching BONE_CONNECTIONS / widget order)
BONE_COLORS = {
    # Upper body
    ("nose", "neck"): OPENPOSE_COLORS[12],
    ("neck", "r_shoulder"): OPENPOSE_COLORS[5],   # Light green - SWAPPED
    ("r_shoulder", "r_elbow"): OPENPOSE_COLORS[6],  # Green - SWAPPED
    ("r_elbow", "r_wrist"): OPENPOSE_COLORS[7],     # Green-cyan - SWAPPED
    ("neck", "l_shoulder"): OPENPOSE_COLORS[1],     # Orange - SWAPPED
    ("l_shoulder", "l_elbow"): OPENPOSE_COLORS[2],  # Dark orange - SWAPPED
    ("l_elbow", "l_wrist"): OPENPOSE_COLORS[3],     # Yellow - SWAPPED
    ("neck", "r_hip"): OPENPOSE_COLORS[11],         # Light blue - SWAPPED
    ("neck", "l_hip"): OPENPOSE_COLORS[8],          # Cyan-green - SWAPPED

    # Right leg
    ("r_hip", "r_knee"): OPENPOSE_COLORS[12],       # Blue - SWAPPED
    ("r_knee", "r_ankle"): OPENPOSE_COLORS[13],     # Purple-blue - SWAPPED

    # Left leg
    ("l_hip", "l_knee"): OPENPOSE_COLORS[9],        # Cyan - SWAPPED
    ("l_knee", "l_ankle"): OPENPOSE_COLORS[10],     # Cyan-blue - SWAPPED
    
    # Face
    ("nose", "r_eye"): OPENPOSE_COLORS[14],
    ("r_eye", "r_ear"): OPENPOSE_COLORS[16],
    ("nose", "l_eye"): OPENPOSE_COLORS[14],
    ("l_eye", "l_ear"): OPENPOSE_COLORS[16],
}

# Joint colors - average of all connected bones or use primary bone color
JOINT_COLORS = {
    # Head
    "nose": OPENPOSE_COLORS[12],       # Blue (nose-neck)
    "neck": OPENPOSE_COLORS[12],       # Blue
    "r_eye": OPENPOSE_COLORS[14],      # Purple
    "l_eye": OPENPOSE_COLORS[14],      # Purple
    "r_ear": OPENPOSE_COLORS[16],      # Pink
    "l_ear": OPENPOSE_COLORS[16],      # Pink
    
    # Right arm (GREEN)
    "r_shoulder": OPENPOSE_COLORS[5],  # Light green
    "r_elbow": OPENPOSE_COLORS[6],     # Green
    "r_wrist": OPENPOSE_COLORS[7],     # Green-cyan
    
    # Left arm (RED-YELLOW)
    "l_shoulder": OPENPOSE_COLORS[1],  # Orange
    "l_elbow": OPENPOSE_COLORS[2],     # Dark orange
    "l_wrist": OPENPOSE_COLORS[3],     # Yellow
    
    # Right leg (BLUE)
    "r_hip": OPENPOSE_COLORS[11],      # Light blue
    "r_knee": OPENPOSE_COLORS[12],     # Blue
    "r_ankle": OPENPOSE_COLORS[13],    # Purple-blue
    
    # Left leg (CYAN)
    "l_hip": OPENPOSE_COLORS[8],       # Cyan-green
    "l_knee": OPENPOSE_COLORS[9],      # Cyan
    "l_ankle": OPENPOSE_COLORS[10],    # Cyan-blue
}

def get_joint_color(joint_name):
    """Get RGB color for a joint.
    
    Args:
        joint_name: Name of the joint
        
    Returns:
        Tuple of (R, G, B) values (0-255)
    """
    return JOINT_COLORS.get(joint_name, (255, 255, 255))

# Fallback palette is unreliable because BONE_CONNECTIONS order varies.
# We will rely on BONE_COLORS lookup.

def get_bone_color(joint1, joint2, bone_index=None):
    """Get RGB color for a bone connection.
    
    Args:
        joint1: Name of first joint
        joint2: Name of second joint
        bone_index: Index in BONE_CONNECTIONS (primary source)
        
    Returns:
        Tuple of (R, G, B) values (0-255)
    """
    # Primary: use bone_index for consistent palette-based colors
    if bone_index is not None and bone_index < len(FALLBACK_PALETTE):
        return FALLBACK_PALETTE[bone_index]
    
    # Secondary: try direct lookup by joint pair
    key = (joint1, joint2)
    if key in BONE_COLORS:
        return BONE_COLORS[key]
    
    # Tertiary: try reverse lookup
    key_reverse = (joint2, joint1)
    if key_reverse in BONE_COLORS:
        return BONE_COLORS[key_reverse]
    
    # Default to white if not found
    return (255, 255, 255)



def get_bone_color_bgr(joint1, joint2, bone_index=None):
    """Get BGR color for a bone connection (for OpenCV).
    
    Args:
        joint1: Name of first joint
        joint2: Name of second joint
        bone_index: Fallback index if direct lookup fails
        
    Returns:
        Tuple of (B, G, R) values (0-255)
    """
    r, g, b = get_bone_color(joint1, joint2, bone_index)
    return (b, g, r)
